dask/bag/core.py

```diff
@@ -1566,62 +1569,39 @@ def to_delayed(self, optimize_graph=True):
+        """ Repartition Bag across new divisions.
 
-        This can be used to reduce or increase the number of partitions
-        of the bag.
+        Parameters
+        ----------
+        npartitions : int, optional
+            Number of partitions of output.
+        partition_size : int or string, optional
+            Max number of bytes of memory for each partition. Use numbers or
+            strings like 5MB.
+
+            .. warning::
+
+               This keyword argument triggers computation to determine
+               the memory size of each partition, which may be expensive.
+
+        Notes
+        -----
+        Exactly one of `npartitions` or `partition_size` should be specified.
+        A ``ValueError`` will be raised when that is not the case.
 
         Examples
         --------
         >>> b.repartition(5)  # set to have 5 partitions  # doctest: +SKIP
         """
-        new_name = "repartition-%d-%s" % (npartitions, tokenize(self, npartitions))
-        if npartitions == self.npartitions:
-            return self
-        elif npartitions < self.npartitions:
-            ratio = self.npartitions / npartitions
-            new_partitions_boundaries = [
-                int(old_partition_index * ratio)
-                for old_partition_index in range(npartitions + 1)
-            ]
-
-            dsk = {}
-            for new_partition_index in range(npartitions):
-                value = (
-                    list,
-                    (
-                        toolz.concat,
-                        [
-                            (self.name, old_partition_index)
-                            for old_partition_index in range(
-                                new_partitions_boundaries[new_partition_index],
-                                new_partitions_boundaries[new_partition_index + 1],
-                            )
-                        ],
-                    ),
-                )
-                dsk[new_name, new_partition_index] = value
-        else:  # npartitions > self.npartitions
-            ratio = npartitions / self.npartitions
-            split_name = "split-%s" % tokenize(self, npartitions)
-            dsk = {}
-            last = 0
-            j = 0
-            for i in range(self.npartitions):
-                new = last + ratio
-                if i == self.npartitions - 1:
-                    k = npartitions - j
-                else:
-                    k = int(new - last)
-                dsk[(split_name, i)] = (split, (self.name, i), k)
-                for jj in range(k):
-                    dsk[(new_name, j)] = (operator.getitem, (split_name, i), jj)
-                    j += 1
-                last = new
-
-        graph = HighLevelGraph.from_collections(new_name, dsk, dependencies=[self])
-        return Bag(graph, name=new_name, npartitions=npartitions)
+        if sum([npartitions is not None, partition_size is not None]) != 1:
```

Comment:
```suggestion
        if npartitions is not None and partition_size is not None:
```

Reply:
@TomAugspurger I don't think that suggestion actually catches the same things as the current `sum`.

IICU, the current `if` is the not-XOR between `npartitions is not None` and  `partition_size is not None` (letting the it be `True` when either both or neither are defined). While your suggestions only would catch the occasion when both are defined.

A possible shortening would be :
```suggestion
        if (npartitions is None) == (partition_size is None):
```
(using casting `is None` to `bool` with the parentheses, the equal operator (`==`, as that is identical to not-XOR for `bool`) , removing the `not` (as that is unnecessary for XOR operations (as it exist on both sides))

I don't have enough knowledge on the codebase of dask to comment on this PR in general.

Reply:
Yeah that's right - the purpose of that check is to ensure only one of `npartitions` and `partition_size` is set. Checking that both are not none allows for both to be set. @sroets way or `not (bool(npartitions) ^ bool(partition_size))` is another.